### PR TITLE
feat: Extend GraphQL via Hooks

### PR DESCRIPTION
**Pull request by fahimalizain**
Added two new hooks:
- graphql_sdl_dir
- graphql_schema_processors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,4 +4,224 @@ GraphQL API Layer for Frappe Framework
...
+    User(filters: "{ \"full_name\": [\"LIKE\", \"%TEST%\"] }") {
+        name, email
+    }
+}
+```  
+
+## RolePermission integration
+Data is returned based on Read access to the resource
+
+## Generic Mutations: set_value & save_doc
```
Comment (assemmarwan) [inline README.md]: @faztp12 If you can please add an example of save_doc

File: README.md [hunk 2]
```diff
@@ -4,4 +4,224 @@ GraphQL API Layer for Frappe Framework
...
+Specify a list of directories containing `.graphql` files relative to the app's root directory.
+eg:
+```py
+# hooks.py
+graphql_sdl_dir = [
+    "./your-app/your-app/generated/sdl/dir1",
+    "./your-app/your-app/generated/sdl/dir2",
+]
+```
+The above will look for graphql files in `your-bench/apps/your-app/your-app/generated/sdl1` folder.
```
Comment (assemmarwan) [inline README.md]: @faztp12 This will also add `dir2` no?

File: README.md [hunk 3]
```diff
@@ -4,4 +4,224 @@ GraphQL API Layer for Frappe Framework
...
+- `graphql_schema_processors`  
+You can pass in a list of cmd that gets executed on schema creation. You are given `GraphQLSchema` object (please refer [graphql-core](https://github.com/graphql-python/graphql-core)) as the only parameter. You can modify it or extend it as per your requirements.
+This is a good place to attach the resolvers for the custom SDLs defined via `graphql_sdl_dir`
+
+### Examples
```
Comment (assemmarwan) [inline README.md]:
```suggestion
## Examples
```

File: README.md [hunk 4]
```diff
@@ -4,4 +4,224 @@ GraphQL API Layer for Frappe Framework
...
+### Introducing a Custom Field to GraphQL
+- Add the `Custom Field` in frappe
+- Add the following to a `.graphql` file in your app and specify its directory via `graphql_sdl_dir`
+```graphql
+extend type User {
+    is_super: Int!
+}
+```
+
+### Adding a Hello World Query
```
Comment (assemmarwan) [inline README.md:254]: @faztp12 If you can please add an example of how the query can be requested for custom queries

File: README.md [hunk 5]
```diff
@@ -4,4 +4,224 @@ GraphQL API Layer for Frappe Framework
...
+The above will look for graphql files in `your-bench/apps/your-app/your-app/generated/sdl1` folder.
+- `graphql_schema_processors`  
+You can pass in a list of cmd that gets executed on schema creation. You are given `GraphQLSchema` object (please refer [graphql-core](https://github.com/graphql-python/graphql-core)) as the only parameter. You can modify it or extend it as per your requirements.
+This is a good place to attach the resolvers for the custom SDLs defined via `graphql_sdl_dir`
+
+### Examples
+### Adding a newly created DocType
+- Generate the SDLs in your app directory
+```bash
+$ bench --site test_site graphql generate_sdl --output-dir <your-app/graphql> --doctype <name>
```
Comment (assemmarwan) [inline README.md]: @faztp12 These arguments aren't defined under `commands/__init__.py` 😅

**Comment by assemmarwan** [comment]
@faztp12 I think we should handle `Link` fields differently, instead of being a `String` they should be the type of that field, this way a single query can be used to get related fields.

For instance,

```graphql
type ToDo implements BaseDocType {
....
  role: String
  assigned_by: User.    <--------
  assigned_by_full_name: String
  sender: String
.....
}
```

By doing this, we can make a query like:

```graphql
{
ToDo {
    name,
    priority,
    assigned_by {
        email,
        full_name
        }
    }
}
```

Thoughts? 🤔

**Comment by fahimalizain** [comment]
> @faztp12 I think we should handle `Link` fields differently, instead of being a `String` they should be the type of that field, this way a single query can be used to get related fields.

@assemmarwan yes, this would be great 👍 

I am wondering if the following can be made possible:
### Query 1:
```graphql
{
  ToDo {
    name,
    assigned_by
  }
}
```
```js
{
  ToDo: [
    { name: "A", assigned_by: "Test User A " },
    ...
  ]
}
```

### Query 2
```graphql
{
  ToDo {
    name,
    assigned_by { full_name, name }
  }
}
```
```js
{
  ToDo: [
    { name: "A", assigned_by: { full_name: "Test A", name: "Test User A" } },
    ...
  ]
}
```

But then, graphql do not support resolving a scalar / object for a single field.
As an alternative, what do you think of this:
image

File: README.md (hunk 1 above)
Comment (fahimalizain) [reply README.md]: @assemmarwan done 👍

File: README.md (hunk 2 above)
Comment (fahimalizain) [reply README.md]: @assemmarwan yep, updated 👍

File: README.md (hunk 3 above)
Comment (fahimalizain) [reply README.md]: updated 👍

**Comment by assemmarwan** [comment]
@faztp12 Currently for the query, if there's a permission error, it returns empty list instead of error. I think it should reflect the errors in the backend

File: README.md (hunk 4 above)
Comment (fahimalizain) [reply README.md:254]: @assemmarwan  looks good ?

Comment (assemmarwan) [reply README.md:254]: @faztp12 Looks good, but I reckon adding a query with variables would make it comprehensive 💯 , what do you think?

Comment (fahimalizain) [reply README.md:254]: @assemmarwan updated and pushed, please check ;)

Comment (assemmarwan) [reply README.md:254]: Looks good 👍🏻

File: README.md (hunk 5 above)
Comment (fahimalizain) [reply README.md]: @assemmarwan updated, please check 👍

**Review by assemmarwan — APPROVED** [review]

**Pull request merged**